atomix/cluster/src/test/java/io/atomix/raft/protocol/ControllableRaftServerProtocol.java

```diff
@@ -0,0 +1,423 @@
+      final MemberId memberId,
+      final Runnable requestHandler,
+      final CompletableFuture responseFuture) {
+    final var message = new Tuple<>(requestHandler, responseFuture);
+    messageQueue.computeIfAbsent(memberId, m -> new LinkedList<>()).add(message);
+  }
+
+  @Override
+  public CompletableFuture<JoinResponse> join(final MemberId memberId, final JoinRequest request) {
+    final var responseFuture = new CompletableFuture<JoinResponse>();
+    send(
+        memberId,
+        () ->
+            getServer(memberId)
+                .thenCompose(listener -> listener.join(request))
+                .thenAccept(
+                    response -> send(localMemberId, () -> responseFuture.complete(response), null)),
+        responseFuture);
+    return responseFuture;
+  }
+
+  @Override
+  public CompletableFuture<LeaveResponse> leave(
+      final MemberId memberId, final LeaveRequest request) {
+    final var responseFuture = new CompletableFuture<LeaveResponse>();
+    send(
+        memberId,
+        () ->
+            getServer(memberId)
+                .thenCompose(listener -> listener.leave(request))
+                .thenAccept(
+                    response -> send(localMemberId, () -> responseFuture.complete(response), null)),
+        responseFuture);
+    return responseFuture;
+  }
+
+  @Override
+  public CompletableFuture<ConfigureResponse> configure(
+      final MemberId memberId, final ConfigureRequest request) {
+    final var responseFuture = new CompletableFuture<ConfigureResponse>();
+    send(
+        memberId,
+        () ->
+            getServer(memberId)
+                .thenCompose(listener -> listener.configure(request))
+                .thenAccept(
+                    response -> send(localMemberId, () -> responseFuture.complete(response), null)),
+        responseFuture);
+    return responseFuture;
+  }
+
+  @Override
+  public CompletableFuture<ReconfigureResponse> reconfigure(
+      final MemberId memberId, final ReconfigureRequest request) {
+    final var responseFuture = new CompletableFuture<ReconfigureResponse>();
+    send(
+        memberId,
+        () ->
+            getServer(memberId)
+                .thenCompose(listener -> listener.reconfigure(request))
+                .thenAccept(
+                    response -> send(localMemberId, () -> responseFuture.complete(response), null)),
+        responseFuture);
+    return responseFuture;
+  }
+
+  @Override
+  public CompletableFuture<InstallResponse> install(
+      final MemberId memberId, final InstallRequest request) {
+    final var responseFuture = new CompletableFuture<InstallResponse>();
+    send(
+        memberId,
+        () ->
+            getServer(memberId)
+                .thenCompose(listener -> listener.install(request))
+                .thenAccept(
+                    response -> send(localMemberId, () -> responseFuture.complete(response), null)),
+        responseFuture);
+    return responseFuture;
+  }
+
```

Comment:
Do you think it is problematic that we have re-implemented the network stuff?

Reply:
I have the feeling we need unit tests for that :sweat_smile:

Reply:
:smile:  Well. Yes, it is complex. It is adapted from the existing TestRaftServerProtocol, which is also simulating the network stuff. We have to control the network messages. So we have to simulate or mock network somehow. Do you think we can do it in a different way?

Reply:
Maybe the real implementation ? I think it has some logic for local right?

Reply:
With real implementation we won't be able to control when a message is delivered or dropped.